src/NServiceBus.Core/Routing/DistributionStrategy.cs

```diff
@@ -24,6 +27,14 @@ protected DistributionStrategy(string endpoint, DistributionStrategyScope scope)
         public abstract string SelectReceiver(string[] receiverAddresses);
 
         /// <summary>
+        /// Selects a destination instance for a message from all known addresses of a logical endpoint.
+        /// </summary>
+        public virtual Task<string> SelectReceiver(string[] receiverAddresses, IOutgoingContext outgoingContext)
```

Comment:
@timbussmann your concerns are absolutely valid, but unfortunately, by comparing it to FileBased routing or instance mapping file, we are comparing apples with oranges. In the case of service fabric, we need to map the payload to a partition range and then the partition range to a partition id. Even though it is valid to assume the number of partitions and therefore the ids to do not change (currently it would mean a destroy of the data) the range to partition id map is always a function over a payload or header. That function can only be executed at runtime and then has to be mapped. Since ranges can be named ranges or Int64 ranges, it is not useful to precompute all possible combinations and cache it. Therefore we have to call into the state manager and resolve it.

Arguably we can move this call to the pipeline in a behavior. But then we have two other problems:

- The routing is currently a stage connector, and we cannot hook into it other than extending the strategy. So the behavior would need to be placed before the sender side connector. If we do that we would need to hack the routing table to ask to what endpoint the message type might be mapped before the connector and based on that we would need to map it to a service fabric URI. So essentially we would take a decision which might be overruled by a routing rule in the connector later. For me that is a no go.
- When we do it in a behavior we still need that call so whether we do it in the strategy or inside the behavior doesn't matter from a latency perspective

I see two options:

- Either we make the strategy async or
- We introduce in a non-breaking way proper routing stages

The latter is much more involved, so that's why I proposed the first. Arguably the default impl always returns an already completed task to there is no transitioning away from the sync path happening and the only overhead we have here is an allocation more.